### ISSUE TITLE
tfpdf needs updating

### ISSUE THREAD

**Issue by rndevuk** [boxbilling/boxbilling#633]
Downloading a pdf invoice for example produces this error/notice: 
```Deprecated: Methods with the same name as their class will not be constructors in a future version of PHP; tFPDF has a deprecated constructor in /src/bb-vendor/rev42/tfpdf/src/tFPDF.php on line 13 Deprecated: Methods with the same name as their class will not be constructors in a future version of PHP; TTFontFile has a deprecated constructor in /src/bb-vendor/rev42/tfpdf/src/font/unifont/ttfonts.php on line 38 Deprecated: The each() function is deprecated. This message will be suppressed on further calls in /src/bb-library/PDF_ImageAlpha.php on line 147 FPDF error: Some data has already been output, can't send PDF file```

I'll work on this once timothygwebb has re-pushed his PR for the php/composer update

**Comment by timothygwebb** [comment]
@rndevuk  This may not need to be completed due to the removal of rev42 in my pull request and its replacement that is compliant with php7.4.9 mpdf. https://github.com/mpdf/mpdf

**Comment by timothygwebb** [comment]
@rndevuk This is still an issue my pdf correction was reverted.

**Comment by fordnox** [comment]
there are in composer file

```
        "fpdf/fpdf": "^1.82",
        "docnet/tfpdf": "^2.2",
        "setasign/fpdf": "^1.8"
```

which is bad, one should be enough. @timothygwebb fix, as you have introduced this https://github.com/boxbilling/boxbilling/commit/20efebbe03e468253416165e030e828676fb0f23

**Comment by rndevuk** [comment]
https://github.com/boxbilling/boxbilling/blame/master/src/bb-library/tfpdf.php#L1

Missing php opening tag. I'm assuming has something todo with your work on tfpd also.

**Comment by evrifaessa** [comment]
@timothygwebb, still in progress?

**Comment by timothygwebb** [comment]
> @timothygwebb, still in progress?

@evrifaessa I am not actively working on this at this time. I am working on a project for my company setting up a broadcast television station. It's a lot of work and is monopolistic of my time. It is also a paying client. I will pick this up when my other project is completed unless you would like to pickup where I left off implementation of docknet/tfpdf.

**Comment by nicaraohosting** [comment]
I downloaded the version that appears in boxbilling.com and I have this problem to download the invoices in pdf and also error ap to visualize the emails sent to the clients

**Comment by evrifaessa** [comment]
> I downloaded the version that appears in boxbilling.com and I have this problem to download the invoices in pdf and also error ap to visualize the emails sent to the clients

Yeah, that is the problem we're fixing with this issue.

**Comment by OstojaOfficial** [comment]
Any progress on this?

**Comment by timothygwebb** [comment]
No. If you would like to pick it up you are welcome to do so.

**Comment by nicaraohosting** [comment]
a query

if someone could check this and have the bills downloadable how much is the least you would accept for financial aid?

In a group on Facebook someone commented on this and wrote this:

"I don't know if I'm wrong, but I can't find the ttfonts.php file

https://github.com/Setasign/tFPDF/blob/a879b83485c263605cbf98ab2e8fdd4d37ca2707/font/unifont/ttfonts.php

It is from the tFPDF library, which uses boxbilling and it is where the TTFontFile class is. In the boxbilling git it is not. "

**Comment by YukariChiba** [comment]
This issue seems to have not been updated for a long time.

> "I don't know if I'm wrong, but I can't find the ttfonts.php file
> 
> https://github.com/Setasign/tFPDF/blob/a879b83485c263605cbf98ab2e8fdd4d37ca2707/font/unifont/ttfonts.php
> 
> It is from the tFPDF library, which uses boxbilling and it is where the TTFontFile class is. In the boxbilling git it is not. "

I copied `tffonts.php` and `font` directories  to `bb-library` directory, and it seems that this fix works at least at php 7.4.
See https://github.com/YukariChiba/boxbilling/commit/2efc3defb69867eb846a24508191d33a808ad72c

Is there a better solution?

**Comment by BenNottelling** [comment]
> This issue seems to have not been updated for a long time.
> 
> > "I don't know if I'm wrong, but I can't find the ttfonts.php file
> > https://github.com/Setasign/tFPDF/blob/a879b83485c263605cbf98ab2e8fdd4d37ca2707/font/unifont/ttfonts.php
> > It is from the tFPDF library, which uses boxbilling and it is where the TTFontFile class is. In the boxbilling git it is not. "
> 
> I copied `tffonts.php` and `font` directories to `bb-library` directory, and it seems that this fix works at least at php 7.4.
> See [YukariChiba@2efc3de](https://github.com/YukariChiba/boxbilling/commit/2efc3defb69867eb846a24508191d33a808ad72c)
> 
> Is there a better solution?

The correct solution is to update to a new PDF generator, but that's a much larger task. I will see about merging the required `tffonts.php` and the bare minimum of fonts to get this working again for our upcoming release, but what I really want is to use a modern and maintained library for this

**Comment by kdktec** [comment]
Hey, Team, Any updates on this issue. We still have the problem.

**Comment by BenNottelling** [comment]
> Hey, Team, Any updates on this issue. We still have the problem.

Apply the patch from here:

**Comment by kdktec** [comment]
> tffonts.php

I can't find any file called "tffonts.php" but there is a file called "ttfonts.php" in /billing/bb-library/font/unifont are you talking about this file and what we should do is move this file into bb-library? Please let me know.

**Comment by kdktec** [comment]
> > Hey, Team, Any updates on this issue. We still have the problem.
> 
> Apply the patch from here: #927 (accidentally added the comment without the link)

Sorry dear,  try all of it. I move the files  YukariChiba said and deleted the unwanted fonts, but still, the problem is the same. 
Can you please give me step by step guide to fix this issue? Because I read all the comments. Its very confusing and hard to find the final solution to fix this issue.

**Comment by nicaraohosting** [comment]
You are going to upload two files in this path:
/bb-library/font/unifont/

there it should go up: DejaVuSansCondensed.ttf and DejaVuSansCondensed-Bold.ttf

The problem that the pdf comes out well centered

**Comment by nicaraohosting** [comment]
These files are on this site
[[boxbilling / src / bb-library / font / unifont /](https://github.com/boxbilling/boxbilling/tree/658f6a1d6a0e602af032b703c34a55c0932c86f7/src/bb-library/font/unifont)

**Comment by sjpals** [comment]
Hi, so sorry, im not really techy ... I went into filemanager and created the folders font/unifont under bb-library and I uploaded the two font files DejaVuSansCondensed.ttf and DejaVuSansCondensed-Bold.ttf . Im still getting the error message. 

The two font files are in mydomain.com/bb-library/font/unifont is that correct?

**Comment by sjpals** [comment]
> You are going to upload two files in this path: /bb-library/font/unifont/
> 
> there it should go up: DejaVuSansCondensed.ttf and DejaVuSansCondensed-Bold.ttf
> 
> The problem that the pdf comes out well centered

Hi, so sorry, im not really techy ... I went into filemanager and created the folders font/unifont under bb-library and I uploaded the two font files DejaVuSansCondensed.ttf and DejaVuSansCondensed-Bold.ttf . Im still getting the error message.

The two font files are in mydomain.com/bb-library/font/unifont is that correct?

**Comment by kdktec** [comment]
I alos copied tffonts.php and font directories to bb-library directory, and it seems that this fix works at least at php 7.4.
You can get files from here: https://github.com/kdktec/boxbilling/tree/master/src/bb-library
